power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/sparse_lu_solver.hpp

```diff
@@ -328,6 +405,210 @@ template <class Tensor, class RHSVector, class XVector> class SparseLUSolver {
     std::shared_ptr<IdxVector const> row_indptr_;
     std::shared_ptr<IdxVector const> col_indices_;
     std::shared_ptr<IdxVector const> diag_lu_;
+    // cache value for pivot perturbation for the factorize step
+    bool has_pivot_perturbation_{false};
+    double matrix_norm_{};
+    std::optional<std::vector<Tensor>> original_matrix_;
+    // cache value for iterative refinement for the solve step
+    std::optional<std::vector<XVector>> dx_;
+    std::optional<std::vector<RHSVector>> residual_;
+    std::optional<std::vector<RHSVector>> rhs_;
+
+    void solve_with_refinement(std::vector<Tensor> const& data,        // pre-factoirzed data, const ref
+                               BlockPermArray const& block_perm_array, // pre-calculated permutation, const ref
+                               std::vector<RHSVector> const& rhs, std::vector<XVector>& x) {
+        // initialize refinement
+        initialize_refinement(rhs, x);
+        double backward_error{10.0};
+        solve_once(data, block_perm_array, rhs_.value(), x);
+        Idx num_iter{};
+        // iterate until convergence
+        while (backward_error > epsilon_perturbation) {
+            if (num_iter++ == max_iterative_refinement) {
+                throw SparseMatrixError{};
+            }
+            // calculate residual
+            calculate_residual(x);
+            // solve with residual
+            solve_once(data, block_perm_array, residual_.value(), dx_.value());
+            // calculate backward error and then iterate x
+            backward_error = iterate_and_backward_error(x);
+        }
```

Comment:
you don't need the first solve, right? the refinement algorithm already contains the first calculation using `residual_ = rhs_` and `dx_ = x` and the backward error may already be small enough after the first solve.

```suggestion
        double backward_error{std::numeric_limits::infinity()};
        Idx num_iter{};
        // iterate until convergence
        while (backward_error > epsilon_perturbation) {
            if (num_iter++ == max_iterative_refinement) {
                throw SparseMatrixError{};
            }
            // solve with residual
            solve_once(data, block_perm_array, residual_.value(), dx_.value());
            // calculate backward error and then iterate x
            backward_error = iterate_and_backward_error(x);
            // calculate residual
            calculate_residual(x);
        }
```

Reply:
I have tried this way before. The program crashed in many places. I could have spent more time to cover some corner cases.

But I decided to revert back to the original.